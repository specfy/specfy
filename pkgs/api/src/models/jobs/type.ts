import type { Prisma } from '@prisma/client';

export interface JobDeployConfig {
  url: string;
  autoLayout?: boolean;
  hook?: {
    id: string;
    ref: string;
  };
}

export type JobWithOrgProject = Prisma.JobsGetPayload<{
  include: { Org: true; Project: true };
}>;

export type JobCode =
  | 'cant_auth_github'
  | 'failed_to_checkout'
  | 'failed_to_cleanup'
  | 'failed_to_clone'
  | 'failed_to_deploy'
  | 'failed_to_start_github_deployment'
  | 'failed_to_teardown'
  | 'missing_dependencies'
  | 'no_api_key'
  | 'org_not_installed'
  | 'project_not_installed'
  | 'success'
  | 'unknown';
