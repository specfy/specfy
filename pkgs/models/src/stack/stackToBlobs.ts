import { slugify, pick, isDiffObjSimple } from '@specfy/core';
import type { Components } from '@specfy/db';
import type { AnalyserJson } from '@specfy/stack-analyser';
import { tech as techList } from '@specfy/stack-analyser';

import type { ComponentType, DBComponent } from '../components/types.js';
import { getComponentSize } from '../flows/helpers.js';
import type { FlowEdge } from '../flows/types.js';
import type {
  ApiBlobCreate,
  ApiBlobCreateComponent,
  PostUploadRevision,
} from '../revisions';
import { findPerfectMatch, findPrevious } from '../stack/index.js';

import { getTitle } from './helpers.js';
import type { StackToBlobs } from './types.js';

const changing: Array<keyof Components> = [
  'edges',
  'source',
  'sourceName',
  'sourcePath',
  'techId',
  'techs',
  'name',
  'inComponent',
];

/**
 * Prepare blobs to create, update or delete
 */
export function stackToBlobs(
  parsed: AnalyserJson,
  prevs: Components[],
  data: PostUploadRevision['Body']
): StackToBlobs {
  const now = new Date().toISOString();
  const unchanged: string[] = [];
  const source = data.source;

  // Find perfect matches first so we don't confuse them later in rename or mv
  const prevIdUsed: string[] = [];
  const perfectMatches = new Map<string, Components>();
  for (const child of parsed.childs) {
    const perfect = findPerfectMatch({ child, prevs, source });
    if (perfect) {
      prevIdUsed.push(perfect.id);
      perfectMatches.set(child.id, perfect);
    }
  }

  // new/old -> new
  const idsMap: Record<string, string> = {};
  // new -> old
  const idsMapReverse: Record<string, string> = {};
  const blobs: ApiBlobCreateComponent[] = [];
  for (const child of parsed.childs) {
    let prev: Components | null = null;
    if (perfectMatches.has(child.id)) {
      prev = perfectMatches.get(child.id)!;
    } else {
      prev = findPrevious({ child, prevs, source, prevIdUsed });
      if (prev) {
        prevIdUsed.push(prev.id);
      }
    }

    let current: DBComponent;
    if (prev) {
      current = {
        ...(prev as unknown as DBComponent),
        source,
        sourceName: child.name,
        sourcePath: child.path,
        techId: child.tech,
        techs: mergeTechs({ prev: prev.techs, next: child.techs, source }),
      };
      if (prev.inComponent.source === source) {
        current.inComponent = { id: child.inComponent, source };
      }

      // Store changed ids
      idsMapReverse[current.id] = child.id;
      idsMap[child.id] = current.id;
      idsMap[current.id] = current.id;
    } else {
      // TODO: try to place new components without overlapping and without auto layout
      const type: ComponentType = child.tech
        ? techList.indexed[child.tech].type
        : 'service';
      current = {
        id: child.id,
        blobId: null,
        orgId: data.orgId,
        projectId: data.projectId,
        name: child.name,
        slug: slugify(child.name),
        type: type,
        typeId: null,
        display: {
          zIndex: 1,
          pos: { x: 0, y: 0 },
          size: getComponentSize(type, child.name),
        },
        edges: child.edges.map((edge) => {
          return {
            ...edge,
            portSource: 'sr',
            portTarget: 'tl',
            vertices: [],
            source,
          };
        }),
        inComponent: { id: child.inComponent, source },
        description: { type: 'doc', content: [] },
        source: data.source,
        sourceName: child.name,
        sourcePath: child.path,
        techId: child.tech,
        techs: child.techs.map((id) => {
          return { id, source };
        }),
        tags: ['github'],
        show: true,
        createdAt: now,
        updatedAt: now,
      };
      idsMap[current.id] = child.id;
    }

    current.name = getTitle(current.name);

    blobs.push({
      created: !prev,
      deleted: false,
      parentId: prev ? prev.blobId : null,
      type: 'component',
      typeId: current.id,
      current,
    });
  }

  const deleted: ApiBlobCreate[] = prevs
    .filter((p) => {
      return !prevIdUsed.includes(p.id) && p.source === source;
    })
    .map((prev) => {
      return {
        created: false,
        deleted: true,
        parentId: prev.blobId,
        type: 'component',
        typeId: prev.id,
        current: prev as any,
      };
    });

  // ---- Apply post compilation change (e.g: deleted hosts, deleted edges, etc.)
  blobs.forEach((blob) => {
    if (blob.deleted) {
      return;
    }

    if (
      blob.current.inComponent.id &&
      blob.current.inComponent.source === source
    ) {
      if (idsMap[blob.current.inComponent.id] === undefined) {
        // Host was deleted
        blob.current.inComponent = { id: null, source };
      } else {
        blob.current.inComponent = {
          id: idsMap[blob.current.inComponent.id],
          source,
        };
      }
    }

    if (!blob.created) {
      const next = parsed.childs.find(
        (child) => child.id === idsMapReverse[blob.typeId]
      )!.edges;
      blob.current.edges = mergeEdges({
        prev: blob.current.edges,
        next,
        source,
        idsMap,
      });
    } else {
      blob.current.edges = mergeEdges({
        prev: [],
        next: blob.current.edges,
        source,
        idsMap,
      });
    }
  });

  // Detect unchanged blobs
  const stats = { created: 0, modified: 0, deleted: 0, unchanged: 0 };
  stats.deleted = deleted.length;
  for (const blob of blobs) {
    if (blob.created) {
      stats.created += 1;
      continue;
    }

    const prev = prevs.find((p) => p.id === blob.typeId)!;
    if (isDiffObjSimple(pick(prev, changing), pick(blob.current, changing))) {
      unchanged.push(prev.id);
      stats.unchanged += 1;
    } else {
      stats.modified += 1;
    }
  }

  // Fail safe
  if (
    blobs.length + deleted.length + unchanged.length >
    prevs.length + parsed.childs.length
  ) {
    throw new Error('More output than input');
  }

  return { blobs, deleted, unchanged, stats };
}

/**
 * Merge techs
 */

function mergeTechs({
  prev,
  next,
  source,
}: {
  prev: Components['techs'];
  next: AnalyserJson['techs'];
  source: string;
}): DBComponent['techs'] {
  if (prev.length <= 0) {
    return next.map((id) => ({ id, source }));
  }

  const techs: DBComponent['techs'] = [];
  for (const tech of prev) {
    if (tech.source !== source) {
      // It's not the same source
      techs.push(tech);
      continue;
    }

    const exist = next.find((id) => id === tech.id);
    if (!exist) {
      // Deleted on the repo
      continue;
    }

    // Found before and after
    techs.push(tech);
  }

  // Find new edges
  for (const tech of next) {
    const exist = prev.find((v) => v.id === tech);
    if (exist) {
      // Already handled before
      continue;
    }

    techs.push({
      id: tech,
      source,
    });
  }

  return techs;
}

/**
 * Merge edges from prev to next
 */
function mergeEdges({
  prev,
  next,
  source,
  idsMap,
}: {
  prev: FlowEdge[];
  next: AnalyserJson['edges'];
  source: string;
  idsMap: Record<string, string>;
}): FlowEdge[] {
  const edges: FlowEdge[] = [];

  // Find old edges and deleted edges
  for (const edge of prev) {
    if (edge.source !== source) {
      // It's not the same source
      // even if it could be the same exact edge, the user put it there so leave them the ownership
      edges.push(edge);
      continue;
    }

    const exist = next.find((v) => idsMap[v.target] === edge.target);
    if (!exist) {
      // Deleted on the repo
      continue;
    }

    // Found before and after
    // NB: We don't update the read/write prop
    edges.push(edge);
  }

  // Find new edges
  for (const edge of next) {
    const exist = prev.find((v) => v.target === idsMap[edge.target]);
    if (exist) {
      // Already handled before
      continue;
    }

    edges.push({
      ...edge,
      target: idsMap[edge.target],
      portSource: 'sr',
      portTarget: 'tl',
      vertices: [],
      source,
    });
  }

  return edges;
}
