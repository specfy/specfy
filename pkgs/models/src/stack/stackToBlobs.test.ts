import type { Components, Projects } from '@specfy/db';
import type { AnalyserJson } from '@specfy/stack-analyser';
import { Payload } from '@specfy/stack-analyser';
import { describe, expect, it } from 'vitest';

import { getBlobComponent } from '../components/index.js';
import type { PostUploadRevision } from '../revisions/index.js';

import { stackToBlobs } from './stackToBlobs.js';

function getDefault(): AnalyserJson {
  return new Payload({ name: 'default', folderPath: '/' }).toJson();
}

describe('stackToBlobs', () => {
  it('should output new blobs once and nothing the second', () => {
    // New upload
    const res = stackToBlobs(
      {
        ...getDefault(),
        childs: [
          { ...getDefault(), name: 'a' },
          { ...getDefault(), name: 'b' },
        ],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res.stats).toStrictEqual({
      created: 2,
      deleted: 0,
      modified: 0,
      unchanged: 0,
    });
    expect(res.blobs).toHaveLength(2);
    expect(res.deleted).toHaveLength(0);
    const snap = res.blobs.map((b) => {
      const copy = {
        ...b,
        id: expect.any(String),
        typeId: expect.any(String),
        current: {
          ...b.current,
          id: expect.any(String),
          createdAt: '',
          updatedAt: '',
        },
      };
      return copy;
    });
    expect(snap).toMatchSnapshot();

    // Re-upload
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        childs: [
          { ...getDefault(), name: 'a' },
          { ...getDefault(), name: 'b' },
        ],
      },
      res.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );

    expect(res2.stats).toStrictEqual({
      created: 0,
      deleted: 0,
      modified: 0,
      unchanged: 2,
    });
    expect(res2.blobs).toHaveLength(2);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.blobs[0].typeId).toBe(res.blobs[0].typeId);
    expect(res2.blobs[1].typeId).toBe(res.blobs[1].typeId);
    expect(res2.unchanged).toStrictEqual([
      res2.blobs[0].typeId,
      res2.blobs[1].typeId,
    ]);
  });

  it('should update id/edges.id when one component was already there', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        childs: [up],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );

    // Re-upload
    const reup: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        childs: [
          reup,
          {
            ...getDefault(),
            name: 'b',
            edges: [
              {
                target: reup.id,
                read: true,
                write: true,
              },
            ],
          },
        ],
      },
      first.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].typeId).toBe(up.id);
    expect(res2.blobs[0].current.id).toBe(up.id);
    expect(res2.blobs[1].current.edges[0].target).toBe(up.id);
    expect(res2.unchanged).toStrictEqual([res2.blobs[0].typeId]);
  });

  it('should detect deleted', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        childs: [up],
      },
      [],
      { source: 'github' } as PostUploadRevision['Body']
    );

    // Re-upload
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        childs: [],
      },
      first.blobs.map((b) => b.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.stats).toStrictEqual({
      created: 0,
      deleted: 1,
      modified: 0,
      unchanged: 0,
    });
    expect(res2.blobs).toHaveLength(0);
    expect(res2.deleted).toHaveLength(1);
    expect(res2.deleted[0].typeId).toBe(up.id);
    expect(res2.unchanged).toStrictEqual([]);
  });

  it('should detect deleted edge and component', () => {
    // New upload
    const a = { ...getDefault(), name: 'a' };
    const b = { ...getDefault(), name: 'b' };
    a.edges.push({
      target: b.id,
      read: true,
      write: true,
    });
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );

    // Re-upload, we loose B and the associated edge
    a.edges = [];
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(1);
    expect(res2.blobs[0].current.edges).toStrictEqual([]);
    expect(res2.deleted).toHaveLength(1);
    expect(res2.deleted[0].typeId).toBe(b.id);
    expect(res2.unchanged).toStrictEqual([]);
  });

  it('should not delete different source', () => {
    // New upload
    const up: AnalyserJson = {
      ...getDefault(),
      name: 'coucou',
      tech: 'algolia',
    };
    const res = stackToBlobs(
      {
        ...getDefault(),
        childs: [up],
      },
      [
        {
          ...(getBlobComponent({
            id: 'foobar',
            orgId: 'acme',
          } as Projects) as unknown as Components),
        },
      ],
      { source: 'github' } as PostUploadRevision['Body']
    );

    expect(res.deleted).toHaveLength(0);
  });

  it('should detect loosing an host', () => {
    // New upload
    const a = { ...getDefault(), name: 'a' };
    const b = { ...getDefault(), name: 'b' };
    a.inComponent = b.id;
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );

    // Re-upload, we loose B
    a.inComponent = null;
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(1);
    expect(res2.deleted).toHaveLength(1);
    expect(res2.blobs[0].current.inComponent).toBeNull();
    expect(res2.deleted[0].typeId).toBe(b.id);
    expect(res2.unchanged).toStrictEqual([]);
  });

  it('should detect adding an host', () => {
    // New upload
    const a = { ...getDefault(), name: 'a' };
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );

    // Re-upload, we add B and it's an host of A
    const b = { ...getDefault(), name: 'b' };
    a.inComponent = b.id;
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].current.inComponent).toBe(res2.blobs[1].current.id);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.unchanged).toStrictEqual([]);
  });
});

describe('edge cases', () => {
  it('should not confuse component with the same name', () => {
    // New upload
    const a: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/src/package.json'],
    };
    const b: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/cmd/api/main.go'],
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );
    expect(first.blobs).toHaveLength(2);

    // Upload again no change
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.unchanged).toStrictEqual([a.id, b.id]);
  });

  it('should not confuse component with the same path', () => {
    // New upload
    const b: AnalyserJson = {
      ...getDefault(),
      name: 'b',
      path: ['/package.json'],
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );
    expect(first.blobs).toHaveLength(1);

    // Upload again but a is coming before b
    // It's a new component with the same path, we should avoid detecting a rename
    const a: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/package.json'],
    };
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].created).toBe(true);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.unchanged).toStrictEqual([b.id]);
  });

  it('should handle two rename (same order)', () => {
    // New upload
    const a: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/cmd'],
    };
    const b: AnalyserJson = {
      ...getDefault(),
      name: 'b',
      path: ['/src'],
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );
    expect(first.blobs).toHaveLength(2);

    // Upload again both components are renamed but there are staying in the same folder
    const c: AnalyserJson = {
      ...getDefault(),
      name: 'c',
      path: ['/cmd'],
    };
    const d: AnalyserJson = {
      ...getDefault(),
      name: 'd',
      path: ['/src'],
    };
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [c, d],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].created).toBe(false);
    expect(res2.blobs[0].typeId).toBe(a.id);
    expect(res2.blobs[1].created).toBe(false);
    expect(res2.blobs[1].typeId).toBe(b.id);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.unchanged).toStrictEqual([]);
  });

  it('should handle one folder rename (different order)', () => {
    // New upload
    const a: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/cmd'],
    };
    const b: AnalyserJson = {
      ...getDefault(),
      name: 'b',
      path: ['/src'],
    };
    const first = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [a, b],
      },
      [],
      {
        source: 'github',
      } as PostUploadRevision['Body']
    );
    expect(first.blobs).toHaveLength(2);

    // Upload again a folder is mv, so it will be matched after
    const c: AnalyserJson = {
      ...getDefault(),
      name: 'a',
      path: ['/z'],
    };
    const d: AnalyserJson = {
      ...getDefault(),
      name: 'b',
      path: ['/src'],
    };
    const res2 = stackToBlobs(
      {
        ...getDefault(),
        name: 'coucou',
        tech: 'algolia',
        childs: [d, c],
      },
      first.blobs.map((blob) => blob.current as unknown as Components),
      { source: 'github' } as PostUploadRevision['Body']
    );
    expect(res2.blobs).toHaveLength(2);
    expect(res2.blobs[0].created).toBe(false);
    expect(res2.blobs[0].typeId).toBe(b.id);
    expect(res2.blobs[1].created).toBe(false);
    expect(res2.blobs[1].typeId).toBe(a.id);
    expect(res2.deleted).toHaveLength(0);
    expect(res2.unchanged).toStrictEqual([b.id]);
  });
});

// ('impossible', () => {

// ('should handle two rename (different order)', () => {
//   // New upload
//   const a: AnalyserJson = {
//     ...getDefault(),
//     name: 'a',
//     path: ['/cmd'],
//   };
//   const b: AnalyserJson = {
//     ...getDefault(),
//     name: 'b',
//     path: ['/src'],
//   };
//   const first = stackToBlobs(
//     {
//       ...getDefault(),
//       name: 'coucou',
//       tech: 'algolia',
//       childs: [a, b],
//     },
//     [],
//     {
//       source: 'github',
//     } as PostUploadRevision['Body']
//   );
//   expect(first.blobs).toHaveLength(2);

//   // Upload again both components are renamed but there are staying in the same folder
//   const c: AnalyserJson = {
//     ...getDefault(),
//     name: 'c',
//     path: ['/z'],
//   };
//   const d: AnalyserJson = {
//     ...getDefault(),
//     name: 'd',
//     path: ['/a'],
//   };
//   const res2 = stackToBlobs(
//     {
//       ...getDefault(),
//       name: 'coucou',
//       tech: 'algolia',
//       childs: [c, d],
//     },
//     first.blobs.map((blob) => blob.current as unknown as Components),
//     { source: 'github' } as PostUploadRevision['Body']
//   );
//   console.log(res2);
//   expect(res2.blobs).toHaveLength(2);
//   expect(res2.blobs[0].created).toBe(false);
//   expect(res2.blobs[0].typeId).toBe(b.id);
//   expect(res2.blobs[1].created).toBe(false);
//   expect(res2.blobs[1].typeId).toBe(a.id);
//   expect(res2.deleted).toHaveLength(0);
//   expect(res2.unchanged).toStrictEqual([]);
// });
// })
