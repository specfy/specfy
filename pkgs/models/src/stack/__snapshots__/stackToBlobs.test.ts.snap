// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stackToBlobs > should output new blobs once and nothing the second 1`] = `
[
  {
    "created": true,
    "current": {
      "blobId": null,
      "createdAt": "",
      "description": {
        "content": [],
        "type": "doc",
      },
      "display": {
        "pos": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 40,
          "width": 130,
        },
        "zIndex": 1,
      },
      "edges": [],
      "id": Any<String>,
      "inComponent": {
        "id": null,
        "source": "github",
      },
      "name": "A",
      "orgId": undefined,
      "projectId": undefined,
      "show": true,
      "slug": "a",
      "source": "github",
      "sourceName": "a",
      "sourcePath": [
        "/",
      ],
      "tags": [
        "github",
      ],
      "techId": null,
      "techs": [],
      "type": "service",
      "typeId": null,
      "updatedAt": "",
    },
    "deleted": false,
    "id": Any<String>,
    "parentId": null,
    "type": "component",
    "typeId": Any<String>,
  },
  {
    "created": true,
    "current": {
      "blobId": null,
      "createdAt": "",
      "description": {
        "content": [],
        "type": "doc",
      },
      "display": {
        "pos": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 40,
          "width": 130,
        },
        "zIndex": 1,
      },
      "edges": [],
      "id": Any<String>,
      "inComponent": {
        "id": null,
        "source": "github",
      },
      "name": "B",
      "orgId": undefined,
      "projectId": undefined,
      "show": true,
      "slug": "b",
      "source": "github",
      "sourceName": "b",
      "sourcePath": [
        "/",
      ],
      "tags": [
        "github",
      ],
      "techId": null,
      "techs": [],
      "type": "service",
      "typeId": null,
      "updatedAt": "",
    },
    "deleted": false,
    "id": Any<String>,
    "parentId": null,
    "type": "component",
    "typeId": Any<String>,
  },
]
`;
